File: src/lgremote/tv/buttons.py
"""Botões aceitos pelo socket de input do webOS.

Whitelist proposital: o nome do botão vem da rede (o PWA) e vira um comando na TV.
Aceitar string arbitrária seria entregar o socket de input pra qualquer um na LAN.
"""

from __future__ import annotations

NAVIGATION = ("UP", "DOWN", "LEFT", "RIGHT", "ENTER", "BACK", "EXIT", "HOME", "MENU")
MEDIA = ("PLAY", "PAUSE", "STOP", "REWIND", "FASTFORWARD")
VOLUME = ("VOLUMEUP", "VOLUMEDOWN", "MUTE")
CHANNEL = ("CHANNELUP", "CHANNELDOWN")
COLORS = ("RED", "GREEN", "YELLOW", "BLUE")
DIGITS = tuple(str(digit) for digit in range(10))
EXTRA = ("INFO", "DASH", "ASTERISK", "CC", "GUIDE", "QMENU")

VALID_BUTTONS: frozenset[str] = frozenset(
    NAVIGATION + MEDIA + VOLUME + CHANNEL + COLORS + DIGITS + EXTRA
)


class UnknownButtonError(ValueError):
    """Botão fora da whitelist."""

    def __init__(self, name: str) -> None:
        super().__init__(f"Botão desconhecido: {name!r}")
        self.name = name


def normalize(name: str) -> str:
    """Valida e normaliza o nome de um botão ('enter' -> 'ENTER')."""
    candidate = name.strip().upper()
    if candidate not in VALID_BUTTONS:
        raise UnknownButtonError(name)
    return candidate

File: app/src/tv/buttons.ts
/**
 * Botões aceitos pelo socket de input do webOS.
 *
 * A whitelist continua fazendo sentido no app: o nome do botão agora vem de uma macro
 * do apps.yaml, e um YAML torto não pode virar comando arbitrário na TV.
 */

export const NAVIGATION = [
  "UP", "DOWN", "LEFT", "RIGHT", "ENTER", "BACK", "EXIT", "HOME", "MENU",
] as const;
export const MEDIA = ["PLAY", "PAUSE", "STOP", "REWIND", "FASTFORWARD"] as const;
export const VOLUME = ["VOLUMEUP", "VOLUMEDOWN", "MUTE"] as const;
export const CHANNEL = ["CHANNELUP", "CHANNELDOWN"] as const;
export const COLORS = ["RED", "GREEN", "YELLOW", "BLUE"] as const;
export const DIGITS = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"] as const;
export const EXTRA = ["INFO", "DASH", "ASTERISK", "CC", "GUIDE", "QMENU"] as const;

export const VALID_BUTTONS: ReadonlySet<string> = new Set<string>([
  ...NAVIGATION, ...MEDIA, ...VOLUME, ...CHANNEL, ...COLORS, ...DIGITS, ...EXTRA,
]);

export class UnknownButtonError extends Error {
  readonly button: string;

  constructor(name: string) {
    super(`Botão desconhecido: ${JSON.stringify(name)}`);
    this.name = "UnknownButtonError";
    this.button = name;
  }
}

/** Valida e normaliza o nome de um botão ('enter' -> 'ENTER'). */
export function normalize(name: string): string {
  const candidate = name.trim().toUpperCase();
  if (!VALID_BUTTONS.has(candidate)) {
    throw new UnknownButtonError(name);
  }
  return candidate;
}
